tests/utils/managed_utils/mod.rs

```diff
@@ -0,0 +1,583 @@
+                welcome.clone(),
+                Some(group.public_tree.clone()),
+            )?;
+        }
+        Ok(())
+    }
+
+    /// Distribute a set of messages sent by the sender with identity
+    /// `sender_id` to their intended recipients in group `Group`. This function
+    /// also verifies that all members of that group agree on the public tree.
+    pub fn distribute_to_members(
+        &self,
+        // We need the sender to know a group member that we know can not have
+        // been removed from the group.
+        sender_id: &Vec<u8>,
+        group: &mut Group,
+        messages: Vec<MLSMessage>,
+    ) -> Result<(), ClientError> {
+        let clients = self.clients.borrow();
+        println!("Distributing and processing messages...");
+        // Distribute message to all members.
+        for (index, member_id) in &group.members {
+            println!("Index: {:?}, Id: {:?}", index, member_id);
+            let member = clients.get(member_id).unwrap().borrow();
+            member.receive_messages_for_group(&group.group_id, messages.clone())?;
+        }
+        // Get the current tree and figure out who's still in the group.
+        let sender = clients.get(sender_id).unwrap().borrow();
+        let sender_groups = sender.groups.borrow();
+        let sender_group = sender_groups.get(&group.group_id).unwrap();
+        group.members = sender
+            .get_members_of_group(&group.group_id)?
+            .iter()
+            .map(|(index, cred)| (index.clone(), cred.identity().clone()))
+            .collect();
+        group.public_tree = sender_group.export_ratchet_tree();
+        drop(sender_group);
+        drop(sender_groups);
+        println!("Group members after distribution:");
+        // Check that the group states of all members match.
+        for (index, m_id) in &group.members {
+            println!("Id: {:?}, Index: {:?}", m_id, index);
+            let m = clients.get(m_id).unwrap().borrow();
+            let group_states = m.groups.borrow_mut();
+            // Some group members may not have received their welcome messages yet.
+            if let Some(group_state) = group_states.get(&group.group_id) {
+                assert_eq!(group_state.export_ratchet_tree(), group.public_tree);
+            };
+            drop(group_states);
+        }
+        Ok(())
+    }
+
+    /// Get `number_of_members` new members for the given group `group` for use
+    /// with the `add_members` function.
+    pub fn random_new_members_for_group(
+        &self,
+        group: &Group,
+        number_of_members: usize,
+    ) -> Result<Vec<Vec<u8>>, SetupError> {
+        let clients = self.clients.borrow();
+        if number_of_members + group.members.len() >= clients.len() {
+            return Err(SetupError::NotEnoughClients);
+        }
+        let mut new_member_ids = Vec::new();
+        for _ in 0..number_of_members {
+            let new_member_id = clients
```

Comment:
Ugh... So you want to move some of the closures here out to functions with names?

Reply:
Right, this is very ugly, but I didn't really know how to do it better. I think moving out the closures won't do that much. Maybe it would be worth using `HashSet`s here?

Reply:
Done!